File: dpb-mcp-go/pkg/analyzer/psr4.go
package analyzer

import (
	"bufio"
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync"

	"github.com/faithfm/php-dependency-mcp/pkg/composer"
	"github.com/faithfm/php-dependency-mcp/pkg/types"
	"golang.org/x/sync/errgroup"
)

// PSR4AnalysisResult represents PSR-4 analysis output
type PSR4AnalysisResult struct {
	Mappings   []types.PSR4Mapping   `json:"mappings"`
	Violations []types.PSR4Violation `json:"violations"`
	Stats      PSR4Stats             `json:"stats"`
}

// PSR4Stats represents PSR-4 statistics
type PSR4Stats struct {
	TotalMappings   int `json:"totalMappings"`
	TotalFiles      int `json:"totalFiles"`
	ValidFiles      int `json:"validFiles"`
	ViolationCount  int `json:"violationCount"`
}

var namespaceRegex = regexp.MustCompile(`namespace\s+([\w\\]+)\s*;`)

// AnalyzePSR4Autoloading analyzes PSR-4 compliance
func AnalyzePSR4Autoloading(repoPath string) (string, error) {
	composerJSON, err := composer.ReadComposerJSON(repoPath)
	if err != nil {
		return "", err
	}

	mappings := composer.GetPSR4Mappings(composerJSON)

	violations := make([]types.PSR4Violation, 0)
	var violationsMu sync.Mutex
	
	totalFiles := 0
	validFiles := 0
	var statsMu sync.Mutex

	// Process each mapping concurrently
	var g errgroup.Group

	for _, mapping := range mappings {
		for _, relativePath := range mapping.Paths {
			absPath := filepath.Join(repoPath, relativePath)
			mappingCopy := mapping

			g.Go(func() error {
				phpFiles, err := findPHPFiles(absPath)
				if err != nil {
					return nil // Skip if directory doesn't exist
				}

				// Process files concurrently
				var fileWg sync.WaitGroup
				for _, file := range phpFiles {
					fileWg.Add(1)
					go func(f string) {
						defer fileWg.Done()

						statsMu.Lock()
						totalFiles++
						statsMu.Unlock()

						namespace, err := extractNamespace(f)
						if err != nil {
							return
						}

						relToRoot, _ := filepath.Rel(absPath, f)
						expectedNS := composer.CalculateExpectedNamespace(mappingCopy.Namespace, relToRoot)

						if namespace == expectedNS {
							statsMu.Lock()
							validFiles++
							statsMu.Unlock()
						} else {
							issue := "Namespace mismatch"
							if namespace == "" {
								issue = "Missing namespace declaration"
							}

							violationsMu.Lock()
							violations = append(violations, types.PSR4Violation{
								File:              filepath.Join(relativePath, relToRoot),
								ExpectedNamespace: expectedNS,
								ActualNamespace:   &namespace,
								Issue:             issue,
							})
							violationsMu.Unlock()
						}
					}(file)
				}
				fileWg.Wait()
				return nil
			})
		}
	}

	g.Wait()

	result := PSR4AnalysisResult{
		Mappings:   mappings,
		Violations: violations,
		Stats: PSR4Stats{
			TotalMappings:  len(mappings),
			TotalFiles:     totalFiles,
			ValidFiles:     validFiles,
			ViolationCount: len(violations),
		},
	}

	jsonData, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		return "", err
	}

	return string(jsonData), nil
}

// findPHPFiles finds all PHP files in a directory
func findPHPFiles(dir string) ([]string, error) {
	files := make([]string, 0)

	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil // Skip errors
		}

		if info.IsDir() {
			name := info.Name()
			if strings.HasPrefix(name, ".") || name == "vendor" || name == "node_modules" {
				return filepath.SkipDir
			}
		}

		if !info.IsDir() && strings.HasSuffix(path, ".php") {
			files = append(files, path)
		}

		return nil
	})

	return files, err
}

// extractNamespace extracts namespace from a PHP file
func extractNamespace(filePath string) (string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return "", err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if matches := namespaceRegex.FindStringSubmatch(line); matches != nil {
			return matches[1], nil
		}
	}

	return "", nil
}

File: dpb-mcp-rust/src/analyzer/psr4.rs
use anyhow::Result;
use lazy_static::lazy_static;
use rayon::prelude::*;
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};
use walkdir::WalkDir;

use crate::composer::{calculate_expected_namespace, get_psr4_mappings, read_composer_json};
use crate::types::{Psr4Mapping, Psr4Violation};

lazy_static! {
    static ref NAMESPACE_RE: Regex = Regex::new(r"namespace\s+([\w\\]+)\s*;").unwrap();
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Psr4AnalysisResult {
    pub mappings: Vec<Psr4Mapping>,
    pub violations: Vec<Psr4Violation>,
    pub stats: Psr4Stats,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Psr4Stats {
    #[serde(rename = "totalMappings")]
    pub total_mappings: usize,
    #[serde(rename = "totalFiles")]
    pub total_files: usize,
    #[serde(rename = "validFiles")]
    pub valid_files: usize,
    #[serde(rename = "violationCount")]
    pub violation_count: usize,
}

pub fn analyze_psr4_autoloading<P: AsRef<Path> + Sync>(repo_path: P) -> Result<String> {
    let composer_json = read_composer_json(&repo_path)?;
    let mappings = get_psr4_mappings(&composer_json);

    let violations = Arc::new(Mutex::new(Vec::new()));
    let total_files = Arc::new(Mutex::new(0usize));
    let valid_files = Arc::new(Mutex::new(0usize));

    // Process each mapping in parallel
    mappings.par_iter().for_each(|mapping| {
        for relative_path in &mapping.paths {
            let abs_path = repo_path.as_ref().join(relative_path);

            if let Ok(php_files) = find_php_files(&abs_path) {
                // Process files in parallel
                php_files.par_iter().for_each(|file| {
                    {
                        let mut count = total_files.lock().unwrap();
                        *count += 1;
                    }

                    if let Ok(namespace) = extract_namespace(file) {
                        if let Ok(rel_to_root) = file.strip_prefix(&abs_path) {
                            let expected_ns = calculate_expected_namespace(
                                &mapping.namespace,
                                &rel_to_root.to_string_lossy(),
                            );

                            if namespace == expected_ns {
                                let mut count = valid_files.lock().unwrap();
                                *count += 1;
                            } else {
                                let issue = if namespace.is_empty() {
                                    "Missing namespace declaration"
                                } else {
                                    "Namespace mismatch"
                                };

                                let mut viols = violations.lock().unwrap();
                                viols.push(Psr4Violation {
                                    file: PathBuf::from(relative_path)
                                        .join(rel_to_root)
                                        .to_string_lossy()
                                        .to_string(),
                                    expected_namespace: expected_ns,
                                    actual_namespace: Some(namespace),
                                    issue: issue.to_string(),
                                });
                            }
                        }
                    }
                });
            }
        }
    });

    let violations = Arc::try_unwrap(violations).unwrap().into_inner().unwrap();
    let total_files = *total_files.lock().unwrap();
    let valid_files = *valid_files.lock().unwrap();
    let total_mappings = mappings.len();
    let violation_count = violations.len();

    let result = Psr4AnalysisResult {
        mappings,
        violations,
        stats: Psr4Stats {
            total_mappings,
            total_files,
            valid_files,
            violation_count,
        },
    };

    Ok(serde_json::to_string_pretty(&result)?)
}

pub fn find_php_files(dir: &Path) -> Result<Vec<PathBuf>> {
    let files: Vec<PathBuf> = WalkDir::new(dir)
        .into_iter()
        .filter_entry(|e| {
            let name = e.file_name().to_string_lossy();
            !name.starts_with('.') && name != "vendor" && name != "node_modules"
        })
        .filter_map(|e| e.ok())
        .filter(|e| e.path().extension().and_then(|s| s.to_str()) == Some("php"))
        .map(|e| e.path().to_path_buf())
        .collect();

    Ok(files)
}

fn extract_namespace(file_path: &Path) -> Result<String> {
    let contents = fs::read_to_string(file_path)?;

    for line in contents.lines() {
        if let Some(captures) = NAMESPACE_RE.captures(line) {
            return Ok(captures[1].to_string());
        }
    }

    Ok(String::new())
}
